File: src/com/easy/PowerOfThree_326.java
/**
 * 
 */
package com.easy;

/**
 * @author Chris Lin
 * @version 1.1
 */
public class PowerOfThree_326 {
    // 2019年7月15日
    // Judge divided by 3 can be modulus
    // Time Complexity: O(logn)
    // Space Complexity:O(1)
    // Runtime: 11 ms, faster than 67.66%
    // Memory Usage: 35.8 MB, less than 5.55%
    public boolean isPowerOfThree(int n) {
        if (n <= 0) {
            return false;
        }

        while (n != 1) {
            if (n % 3 != 0) {
                return false;
            }
            n /= 3;
        }
        return true;
    }

    // 2019年7月15日
    // Use Math formula
    // log3(n) = logb(n)/logb(3)
    // Time Complexity: O(logn)
    // Space Complexity:O(1)
    // Runtime: 12 ms, faster than 16.79%
    // Memory Usage: 36 MB, less than 5.01%
    public boolean isPowerOfThree2(int n) {
        return Math.log10(n) / Math.log10(3) % 1 == 0;
    }

    // 2019年9月26日
    // Use Recursive
    // Time Complexity: O(logn)
    // Space Complexity:O(1)
    // Runtime: 11 ms, faster than 78.87%
    // Memory Usage: 35.4 MB, less than 9.38%
    public boolean isPowerOfThree3(int n) {
        return (n > 0) && ((n == 1) || (n % 3 == 0 && isPowerOfThree3(n / 3)));
    }

    // 2019年9月26日
    // Simplify Judge divided by 3 can be modulus
    // Time Complexity: O(logn)
    // Space Complexity:O(1)
    // Runtime: 11 ms, faster than 78.87%
    // Memory Usage: 35.9 MB, less than 6.25%
    public boolean isPowerOfThree4(int n) {
        if (n <= 0) {
            return false;
        }
        while (n % 3 == 0) {
            n /= 3;
        }

        return (n == 1);
    }
}

File: Golang_src/easy/powerOfThree_326.go
package easy

// Use recursive:
// Time Complexity: O(log3n)
// Space Complexity:O(1)
// Runtime: 16 ms, faster than 93.65%
// Memory Usage: 6.1 MB, less than 100.00%
func isPowerOfThree(n int) bool {
	if n <= 0 {
		return false
	}

	for n > 1 {
		if n%3 != 0 {
			return false
		}
		n /= 3
	}
	return true
}

// Use Integer Limitations:
// Time Complexity: O(1)
// Space Complexity:O(1)
// Runtime: 36 ms, faster than 35.71%
// Memory Usage: 6.1 MB, less than 100.00%
func isPowerOfThree2(n int) bool {
	return n > 0 && 1162261467%n == 0
}
